src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshots.java

```diff
@@ -70,17 +123,120 @@ public FileInfo findPhysicalIndexFile(String physicalName) {
-        }
-        return null;
+        return files.get(name);
     }
 
     @Override
-    public Iterator<BlobStoreIndexShardSnapshot> iterator() {
+    public Iterator<SnapshotFiles> iterator() {
         return shardSnapshots.iterator();
     }
+
+    static final class Fields {
+        static final XContentBuilderString FILES = new XContentBuilderString("files");
+        static final XContentBuilderString SNAPSHOTS = new XContentBuilderString("snapshots");
+    }
+
+    static final class ParseFields {
+        static final ParseField FILES = new ParseField("files");
+        static final ParseField SNAPSHOTS = new ParseField("snapshots");
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startArray(Fields.FILES);
+        for (Map.Entry<String, FileInfo> entry : files.entrySet()) {
+            FileInfo.toXContent(entry.getValue(), builder, params);
+        }
+        builder.endArray();
+
+        builder.startObject(Fields.SNAPSHOTS);
+        for (SnapshotFiles snapshot : shardSnapshots) {
+            builder.startObject(snapshot.snapshot(), XContentBuilder.FieldCaseConversion.NONE);
+            builder.startArray(Fields.FILES);
+            for (FileInfo fileInfo : snapshot.indexFiles()) {
+                builder.value(fileInfo.name());
+            }
+            builder.endArray();
+            builder.endObject();
+        }
+        builder.endObject();
+
+        builder.endObject();
+        return builder;
+    }
+
+    public static BlobStoreIndexShardSnapshots fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = parser.currentToken();
+        Map<String, List<String>> snapshotsMap = newHashMap();
+        ImmutableMap.Builder<String, FileInfo> filesBuilder = ImmutableMap.builder();
+        if (token == XContentParser.Token.START_OBJECT) {
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    String currentFieldName = parser.currentName();
+                    token = parser.nextToken();
+                    if (token == XContentParser.Token.START_ARRAY) {
+                        if (ParseFields.FILES.match(currentFieldName)) {
+                            while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                                FileInfo fileInfo = FileInfo.fromXContent(parser);
+                                filesBuilder.put(fileInfo.name(), fileInfo);
+                            }
+                        } else {
+                            throw new ElasticsearchParseException("unknown array [" + currentFieldName + "]");
+                        }
+                    } else if (token == XContentParser.Token.START_OBJECT) {
+                        if (ParseFields.SNAPSHOTS.match(currentFieldName)) {
```

Comment:
Or we can let poorly formatted values pass through and throw exceptions at the end if values are missing, similar to how we do for queries